streamparse/dsl/component.py

```diff
@@ -1,30 +1,177 @@
 """
 Component-level Specification
 
-This module is called component to mirror organization of storm package.
+This module is called component to mirror organization of pystorm package.
 """
+from __future__ import absolute_import
 
-from pystorm.component import Component
+from copy import deepcopy
 
+import simplejson as json
+from six import iteritems, string_types
 
-class Specification(object):
-    def __init__(self, component_cls, name=None, parallelism=1):
-        if not issubclass(component_cls, Component):
-            raise TypeError("Invalid component: {}".format(component_cls))
+from .stream import Grouping, Stream
+from .thrift import storm_thrift
+from storm_thrift import (ComponentCommon, ComponentObject, GlobalStreamId,
+                          JavaObject, ShellComponent, StreamInfo)
 
-        if not isinstance(parallelism, int) or parallelism < 1:
+
+class ComponentSpec(object):
+    def __init__(self, component_cls, name=None, inputs=None, parallelism=1,
+                 config=None, outputs=None):
+        # Grab class attribute versions of args if necessary
+        if parallelism is None:
+            parallelism = component_cls.parallelism
+        if config is None:
+            config = component_cls.config
+        if outputs is None:
+            config = component_cls.outputs
+        # Validate args
+        if not isinstance(parallelism, int):
+            raise TypeError("Parallelism must be a integer greater than 0")
+        elif parallelism < 1:
             raise ValueError("Parallelism must be a integer greater than 0")
+        # if not issubclass(component_cls, Component):
+        #     raise TypeError("Invalid component: {}".format(component_cls))
+        if isinstance(inputs, dict):
```

Comment:
the `inputs` validation looks like it should be its own sub-method, and perhaps you could use [singledispatch](https://pypi.python.org/pypi/singledispatch) to clarify the if/else chain by breaking it into several methods each bound to a different type for dispatch. Just a thought for refactoring.